ARROW-17288: [C++] Adapt the CSV file format to the new scan API

westonpace opened the pull request:
This also adds a utility method arrow::compute::DeclarationToReader.

This does not fully satisfy ARROW-17288 as it only adapts the CSV file format.  However, this was the most complicated case, and I had to convert some of the common test utilities as well.  So I am hopeful the other formats will be more straightforward.

rok reviewed (commented):
Thanks for doing this Weston!

It's a little bit unclear to me what happens type-wise when csv column names are duplicated and types of duplicated columns don't match. Do we currently raise or promote?

Otherwise I'm not that familiar with the codebase so just some naming comments.

westonpace commented:
> It's a little bit unclear to me what happens type-wise when csv column names are duplicated and types of duplicated columns don't match. Do we currently raise or promote?

This should be fully configurable by the schema evolution strategy.

* The user will ask for a column in the dataset schema and we will know the name of that column
* The CSV reader will supply its list of column names (which will contain the duplicate)
* The evolution strategy will be given that list (the inspected fragment), the dataset schema, and the requested columns in the dataset schema and must tell us which columns to load from the CSV.

I think the default behavior picks one arbitrarily and tries to load it.  If the type for that column in the file doesn't match the type for the requested column in the dataset schema then a runtime error should be generated.

Happy to change the default to simply give an error when the fragment has duplicate column names.  At this point in time I don't think we actually infer the types but we could, so that we could support a "find first column with the right name and a compatible type".
In most cases, if the fragments have duplicate column names, any amount of guessing on are part isn't likely to be correct.

Sounds like a good test case to have though to verify whatever it is that we do.

rok commented:
> I think the default behavior picks one arbitrarily and tries to load it. If the type for that column in the file doesn't match the type for the requested column in the dataset schema then a runtime error should be generated.

That sounds ok correctness wise, but also like it could make execution non-deterministic for preventable reasons?

lidavidm reviewed (commented):
Overall looks good to me.

lidavidm commented:
I also lean towards having an explicit error if the column is ambiguous. Trying to type-match also feels somewhat brittle to me.

westonpace commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -178,6 +178,13 @@ struct ARROW_DS_EXPORT ScanV2Options : public compute::ExecNodeOptions {
+  /// The scan node will not automatically fetch fields referenced by the filter
+  /// expression. \see AddFieldsNeededForFilter
+  ///
+  /// If the filter references fields that are not included in `columns` this may or may
+  /// not be an error, depending on the format.
```
Yes, formats like CSV and IPC which just ignore the filter will not error.  Even using parquet as an example, one could filter on statistics for a column, without ever actually loading the column.  That's generally not what is desired, since that filter is best-effort, and the column would then be needed to fulfill the filter in-memory.

westonpace commented on cpp/src/arrow/dataset/scanner.h:
```diff
@@ -243,10 +250,17 @@ struct ARROW_DS_EXPORT ScanV2Options : public compute::ExecNodeOptions {
+  /// This method adds any fields that are needed by `filter` which are not already
+  /// included in the list of columns.  Any new fields added will be added to the end
+  /// in no particular order.
+  static Status AddFieldsNeededForFilter(ScanV2Options* options);
```
There is a (potentially rare) possibility that a format could fully satisfy the best effort filter.  In that case, if the field is only needed for filtering, we shouldn't load it.

Although, in that case, @jacques-n has convinced me that we might want two filter fields, "best-effort filter" which does not have to be satisfied (and thus should always be a selected column) and `filter` which MUST be satisfied (and the format should error if it cannot, which doesn't have to be in the columns list.

Either way, the main reason I'm not automagically including this is because I am trying to move away from the concept of a scan node automagically adding more columns than the user asked for.  I will be doing something similar in ARROW-16072 so that the __filename, etc. fields are only included if asked for.

The goal is that, a scan request with X columns will emit batches with X columns.

> Why isn't this just a regular method?

I could be convinced otherwise.  In my head all exec node options objects are PODs and I never know whether a POD should have methods or not.  I think, however, it would be fine to make an immutable 

```
Result<ScanV2Options> AddFIeldsNeededForFilter() const
```

Would that be preferred?

lidavidm replied:
Oops, sorry, I missed this. The reasoning makes sense. I suppose if it's meant to be an immutable POD, then the latter method makes more sense to me. But that's kind of a nit.

westonpace commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -108,6 +115,9 @@ class ARROW_DS_EXPORT FileSource : public util::EqualityComparable<FileSource> {
   /// \brief Get a RandomAccessFile which views this file source
   Result<std::shared_ptr<io::RandomAccessFile>> Open() const;
 
+  /// \brief Get the size (in bytes) of the file or buffer
```
Hmmm...the same question exists for "encoded size?" or "decoded size?".  I believe the answer is "uncompressed and decoded".  The goal is to limit how much data is loaded after a scanner is paused.  I will update this comment.

westonpace replied:
Ah, I misunderstood and thought we were talking about a different size.  This size could pose a problem for compressed streams.  It should be the compressed size since this is referring to the size of the file.  I'll have to think over the implications (or test) of compressed streams a bit more.

westonpace commented on cpp/src/arrow/dataset/scanner.cc:
```diff
@@ -68,14 +68,28 @@ std::vector<FieldRef> ScanOptions::MaterializedFields() const {
+  for (const auto& field : fields_referenced) {
+    // Note: this will fail if the field reference is ambiguous or the field doesn't
+    // exist in the dataset schema
+    ARROW_ASSIGN_OR_RAISE(auto field_path, field.FindOne(*options->dataset->schema()));
```
ARROW-18387

westonpace commented:
> I also lean towards having an explicit error if the column is ambiguous. Trying to type-match also feels somewhat brittle to me.

There was a slightly related discussion here: https://github.com/apache/arrow/pull/13938#issuecomment-1224349673 where the following was suggested:

> Hmm, it would be nice if this could work even without disabling schema evolution. Perhaps a heuristic is possible?
> 
>  *  if a field name is unique, do as usual
>  *  if a field name is non-unique, require that it has the same number of occurrences in both schema, and iterate on those pairs in order

lidavidm commented:
That also seems reasonable. As long as the behavior is well-specified.

lidavidm commented:
It's also something that could be left for a future PR, IMO

westonpace commented:
> It's also something that could be left for a future PR, IMO

ARROW-18388

westonpace commented on cpp/src/arrow/compute/exec/exec_plan.cc:
```diff
@@ -614,6 +614,106 @@ Result<std::vector<ExecBatch>> DeclarationToExecBatches(Declaration declaration,
+        });
+  }
+
+  // TODO: Should use exec context owned by the plan
```
Ah, no it does not need to be addressed any longer.  The plan now uses the context of the batch converter so there is a single exec context, which was what the todo comment was referring to.

westonpace commented on cpp/src/arrow/compute/exec/exec_plan.cc:
```diff
@@ -614,6 +614,106 @@ Result<std::vector<ExecBatch>> DeclarationToExecBatches(Declaration declaration,
+Result<std::unique_ptr<RecordBatchReader>> DeclarationToReader(Declaration declaration,
+                                                               bool use_threads) {
+  std::shared_ptr<Schema> schema;
+  auto batch_itr = std::make_unique<Iterator<std::shared_ptr<RecordBatch>>>(
```
Added, thanks.

westonpace commented on cpp/src/arrow/compute/exec/exec_plan.cc:
```diff
@@ -614,6 +614,106 @@ Result<std::vector<ExecBatch>> DeclarationToExecBatches(Declaration declaration,
+    std::unique_ptr<Iterator<std::shared_ptr<RecordBatch>>> iterator_;
+  };
+
+  return std::make_unique<PlanReader>(std::move(schema), std::move(batch_itr));
```
Added

westonpace commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -108,6 +131,18 @@ Future<std::optional<int64_t>> FileFormat::CountRows(
+Future<std::shared_ptr<InspectedFragment>> FileFormat::InspectFragment(
+    const FileSource& source, const FragmentScanOptions* format_options,
+    compute::ExecContext* exec_context) const {
+  return Status::NotImplemented("The new scanner is not yet supported on this format");
```
I changed this around to read "scan2 node".  Although I hope we don't end up needing a scan2 node and this can just replace the scan node.

westonpace commented on cpp/src/arrow/dataset/file_csv.cc:
```diff
@@ -52,9 +53,99 @@ using internal::SerialExecutor;
+        best_guess_bytes_per_batch_(best_guess_bytes_per_batch) {}
+
+  Future<std::shared_ptr<RecordBatch>> ScanBatch(int batch_number) override {
+    // We should be called in increasing order but let's verify that in case it changes.
```
Yes.  Thanks.  Fixed.

westonpace commented on cpp/src/arrow/compute/exec/expression.h:
```diff
@@ -220,6 +220,8 @@ ARROW_EXPORT
 Result<Expression> SimplifyWithGuarantee(Expression,
                                          const Expression& guaranteed_true_predicate);
 
+ARROW_EXPORT Result<Expression> RemoveNamedRefs(Expression expression);
```
Added a docstring and a few basic tests (`Expression.RemoveNamedRefs`)

westonpace commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -175,6 +177,7 @@ class ARROW_EXPORT SinkNodeOptions : public ExecNodeOptions {
   /// data from the plan.  If this function is not called frequently enough then the sink
   /// node will start to accumulate data and may apply backpressure.
   std::function<Future<std::optional<ExecBatch>>()>* generator;
+  std::shared_ptr<Schema>* schema;
```
Added a docstring

lidavidm commented on cpp/src/arrow/dataset/file_csv.cc:
```diff
@@ -52,9 +53,99 @@ using internal::SerialExecutor;
+    // It would be easy enough to handle out of order but no need for that complexity at
+    // the moment.
+    DCHECK_EQ(scanned_so_far_++, batch_number);
+    return reader_->ReadNextAsync().Then(
+        [](const std::shared_ptr<RecordBatch>& batch) { return batch; });
```
Hmm, isn't the `Then` here redundant?

westonpace replied:
Good catch.

pitrou commented on cpp/src/arrow/compute/exec/expression_test.cc:
```diff
@@ -927,6 +927,26 @@ TEST(Expression, FoldConstantsBoolean) {
+
+    EXPECT_EQ(without_named_refs, expected);
+  }
+} ExpectRemovesRefsTo;
```
Am I missing something that makes this function better/more useful than a simple function?

westonpace replied:
No, I was copying `ExpectFoldsTo` but I see there are other `ExpectsXyz` methods in this file which aren't structs and so I changed it to a plain method for simplicity.

pitrou commented on cpp/src/arrow/compute/exec/expression.cc:
```diff
@@ -1191,6 +1191,26 @@ Result<Expression> SimplifyWithGuarantee(Expression expr,
+  }
+  return ModifyExpression(
+      std::move(src),
+      [](Expression expr) {
```
Nit: add parameter names for these callbacks?
```suggestion
      /*pre=*/ [](Expression expr) {
```

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/compute/exec/expression.cc:
```diff
@@ -1191,6 +1191,26 @@ Result<Expression> SimplifyWithGuarantee(Expression expr,
+
+        return expr;
+      },
+      [](Expression expr, ...) { return expr; });
```
```suggestion
      /*post_call=*/ [](Expression expr, ...) { return expr; });
```

westonpace replied:
Fixed.

pitrou commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -214,9 +214,11 @@ struct ARROW_EXPORT BackpressureOptions {
 class ARROW_EXPORT SinkNodeOptions : public ExecNodeOptions {
  public:
   explicit SinkNodeOptions(std::function<Future<std::optional<ExecBatch>>()>* generator,
```
Do we want to ensure backwards compatibility and still provide the former 3-argument constructor?

westonpace replied:
Good idea.  Done.

pitrou commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -497,5 +497,30 @@ typename Fut::SyncType RunSynchronously(FnOnce<Fut(Executor*)> get_future,
+///   Tasks will be left in a queue until the next call and no work will be done between
+///   calls.
+template <typename T>
+Iterator<T> IterateSynchronously(
```
Can you add basic tests for this?

westonpace replied:
Done.  The underlying IterateGenerator has quite a few tests already so I just tested the basic functionality in this method.

pitrou reviewed (commented):
Some comments while I skim through this. This isn't a comprehensive review at all.

westonpace commented:
I believe I have addressed the review comments and CI is green again.  I will probably merge this soon.  We can still catch / change things as we address the other formats (and hook in the bindings) too.

zifengyu commented on cpp/src/arrow/dataset/file_csv.cc:
```diff
@@ -52,9 +53,98 @@ using internal::SerialExecutor;
+    // This should be called in increasing order but let's verify that in case it changes.
+    // It would be easy enough to handle out of order but no need for that complexity at
+    // the moment.
+    DCHECK_EQ(scanned_so_far_++, batch_number);
```
will ScanNode::ScanBatchTask call ScanBatch unorderly and cause failure?
